state/state.go

```diff
@@ -34,6 +40,110 @@ type DownloadedObject struct {
 	Object, Filename string
 }
 
+type DynamoDBStater struct {
+	*sync.Mutex
+	Session *session.Session
+	Service string
+}
+
+func NewDynamoDBStater(session *session.Session, service string) *DynamoDBStater {
+	return &DynamoDBStater{
+		Mutex:   &sync.Mutex{},
+		Session: session,
+		Service: service,
+	}
+}
+
+// list of processed objects
+func (d *DynamoDBStater) processedObjects() (map[string]time.Time, error) {
+	objs := make(map[string]time.Time)
+
+	type Record struct {
+		S3object string    `json:"S3object"`
+		Time     time.Time `json:"Time"`
+	}
+
+	var records []Record
+
+	svc := dynamodb.New(d.Session)
+	err := svc.ScanPages(&dynamodb.ScanInput{
+		TableName: aws.String("HoneyELBAccessLogBuckets"),
+	}, func(logs *dynamodb.ScanOutput, last bool) bool {
+		recs := []Record{}
+
+		err := dynamodbattribute.UnmarshalListOfMaps(logs.Items, &recs)
+
+		// break out of function
+		if err != nil {
+			panic(fmt.Sprintf("failed to unmarshal Dynamodb Scan Items, %v", err))
+		}
+
+		records = append(records, recs...)
+
+		return true
+	})
+
+	if err != nil {
+		return objs, fmt.Errorf("Error scanning DynamoDB, %v", err)
+	}
+
+	for _, record := range records {
+		objs[record.S3object] = record.Time
+	}
+	return objs, nil
+}
+
+func (d *DynamoDBStater) ProcessedObjects() (map[string]time.Time, error) {
+	d.Lock()
+	defer d.Unlock()
+	return d.processedObjects()
+}
+func (d *DynamoDBStater) SetProcessed(s3object string) error {
+
+	// add access log information
+	svc := dynamodb.New(d.Session)
+
+	type Object struct {
+		S3object string
+		Time     time.Time
+	}
+
+	objMap := Object{
+		S3object: s3object,
+		Time:     time.Now(),
+	}
+
+	obj, err := dynamodbattribute.MarshalMap(objMap)
+
+	if err != nil {
+		return fmt.Errorf("Marshalling DynamoDB object failed: %s", err)
+	}
+
+	// add object to dynamodb using conditional
+	// if the object exists, no write happens
+	input := &dynamodb.PutItemInput{
+		Item:                obj,
+		TableName:           aws.String("HoneyELBAccessLogBuckets"),
+		ConditionExpression: aws.String("attribute_not_exists(S3object)"),
+	}
+
+	_, err = svc.PutItem(input)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			// we want this to happen if object already exists
+			if awsErr.Code() != dynamodb.ErrCodeConditionalCheckFailedException {
+				return fmt.Errorf("PutItem failed: %s", err)
+			}
+		}
+		// if it is the conditional check, we can just pop out and
+		// ignore this object!
+		return nil
+	}
+
```

Comment:
Might want to consider deleting old items here too so the table doesn't grow indefinitely. Only the past ~hour of items will be useful. e.g., in `FileStater`: https://github.com/honeycombio/honeyaws/blob/master/state/state.go#L96-L102

Reply:
edit: TTL will take care of this ✅